### PR TITLE
SAK-48845 Library Improve Portal Theme Switcher

### PR DESCRIPTION
**Pull request by kunaljaykam**
https://sakaiproject.atlassian.net/browse/SAK-48845

### CODE REVIEW

File: library/src/webapp/js/headscripts.js [hunk 1]
```diff
@@ -832,6 +832,33 @@ function includeWebjarLibrary(library) {
...
+/**
+ * Applies a theme if unset, serving as a fallback without portal.theme.switcher.js.
+ */
+function applyThemeIfMissing() {
+
+	const sakaiTheme = localStorage.getItem('sakai-theme');
+	if (sakaiTheme) {
+	  const htmlElement = document.documentElement;
+	  if (!htmlElement.classList.contains(sakaiTheme)) {
+		htmlElement.classList.forEach((className) => {
```
Comment (adrianfish) [inline library/src/webapp/js/headscripts.js:844]: You don't need to add parentheses around single lambda arguments. We don't do that in Java either. The linting rules for the webcomponents will complain about unnecessary parens.

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 2]
```diff
@@ -7,31 +7,36 @@ portal.lightThemeClass = 'sakaiUserTheme-light';
 portal.darkThemeClass = 'sakaiUserTheme-dark';
 
 portal.toggleDarkTheme = () => {
-  // toggle the dark theme switch to the opposite state
   portal.darkThemeSwitcher && portal.darkThemeSwitcher.getAttribute("aria-checked") === "false" ? portal.enableDarkTheme() : portal.enableLightTheme();
 };
 
 portal.setDarkThemeSwitcherToggle = onOff => {
-  portal.darkThemeSwitcher && portal.darkThemeSwitcher.setAttribute("aria-checked", onOff);
+  portal.darkThemeSwitcher && portal.darkThemeSwitcher.setAttribute("aria-checked", onOff.toString());
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:14]: You can use optional chaining here.
portal.darkThemeSwitcher?.setAttribute

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 3]
```diff
@@ -7,31 +7,36 @@ portal.lightThemeClass = 'sakaiUserTheme-light';
...
-  const url = `/direct/userPrefs/updateKey/${portal.user.id}/sakai:portal:theme?theme=${theme}`;
-  fetch(url, { method: "PUT", credentials: "include" })
-  .then(r => {
-
-    if (!r.ok) {
-      throw new Error(`Network error while updating theme pref at ${url}`);
-    }
-  })
-  .catch (error => console.error(error));
+  if (portal.user && portal.user.id) {
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:29]: portal?.user?.id

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 4]
```diff
@@ -7,31 +7,36 @@ portal.lightThemeClass = 'sakaiUserTheme-light';
...
+    const url = `/direct/userPrefs/updateKey/${portal.user.id}/sakai:portal:theme?theme=${theme}`;
+    fetch(url, { method: "PUT", credentials: "include" })
+      .then(r => {
+        if (!r.ok) {
+          throw new Error(`Network error while updating theme pref at ${url}`);
+        }
+        localStorage.setItem('sakai-theme', theme); // save the theme to local storage
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:36]: I don't think you need this comment. It's pretty obvious what's happening :)

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 5]
```diff
@@ -75,42 +78,37 @@ portal.enableDarkTheme = () => {
...
-// if the dark theme switch is on the page, attach listener to dark theme toggle switch
-portal.darkThemeSwitcher && portal.darkThemeSwitcher.addEventListener('click', portal.toggleDarkTheme, false);
-
-if (portal.userThemeAutoDetectDark) {
-  if (portal.user.id) {
-    // only check for unset theme preference because light and dark themes are already set by Java
-    if (portal.userTheme === portal.defaultThemeClass) {
-      // if the user has dark mode set on their OS, enable dark mode
-      portal.isOsDarkThemeSet() ? portal.enableDarkTheme() : portal.setPortalThemeUserPref(portal.lightThemeClass);
+  let themeToApply = (portal.isOsDarkThemeSet() ? portal.darkThemeClass : portal.lightThemeClass);
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:89]:
```suggestion
  let themeToApply = portal.isOsDarkThemeSet() ? portal.darkThemeClass : portal.lightThemeClass;
```

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 6]
```diff
@@ -75,42 +78,37 @@ portal.enableDarkTheme = () => {
...
+  if (!portal.darkThemeSwitcher) {
+    // If the switcher is not available, directly apply the OS theme preference
+    portal.addCssClassToMarkup(themeToApply);
+  } else {
+    // If the user is logged in and theme is not set or if the switcher is available, follow the logic based on user preference or auto-detect
+    if (portal.user && portal.user.id && portal.userTheme === portal.defaultThemeClass) {
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:95]:
```suggestion
    if (portal?.user?.id && portal.userTheme === portal.defaultThemeClass) {
```

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 7]
```diff
@@ -75,42 +78,37 @@ portal.enableDarkTheme = () => {
...
   }
-}
 
-if (document.documentElement.classList.contains(portal.darkThemeClass)) {
-  // the dark theme switch toggle is off by default, so toggle it to on if dark theme is enabled
-  portal.setDarkThemeSwitcherToggle(true);
-}
+  // Ensure the switch toggle state matches the applied theme
+  if (themeToApply === portal.darkThemeClass) {
+
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:109]:
```suggestion
```

File: portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/SiteHandler.java [hunk 8]
```diff
@@ -509,6 +509,13 @@ public void doSite(HttpServletRequest req, HttpServletResponse res, Session sess
 			ResourcePropertiesEdit props = prefs.getPropertiesEdit(org.sakaiproject.user.api.PreferencesService.SITENAV_PREFS_KEY);
 			props.addProperty(PortalConstants.PROP_CURRENT_EXPANDED, "true");
 			props.addProperty(PortalConstants.PROP_EXPANDED_SITE, siteId);
+
+			boolean themeEnabled = ServerConfigurationService.getBoolean("portal.themes", true);
```
Comment (adrianfish) [inline portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/SiteHandler.java:513]: You can use PortalConstants.PROP_PORTAL_THEMES

**Review by adrianfish — CHANGES_REQUESTED** [review]
Seems decent to me, thanks. Just a few small things.

**Comment by kunaljaykam** [comment]
> Seems decent to me, thanks. Just a few small things.

I added all the changes you requested. Thank you!

**Review by adrianfish — APPROVED** [review]

**Pull request merged**